### DISCUSSION TITLE
How to deal with data file corrupted situation

### DISCUSSION THREAD

**Discussion by Reidddddd** [milvus-io/milvus#37497] [Q&A and General discussion] [answered]
One of an insert_log stored in S3 is corrupted, and couldn't recover (all of its blocks are gone)

I observed that the compaction keeps compacting the corrupted insert_log, failed and re-downloaded looping endless..

Want to check what's the proper action to deal with these kinds of situation

**Comment by yhmo** [comment, marked as answer]
The path of the corrupted file is still recorded in etcd, so milvus tries to compact it.

The is a developer tool birdwatch: https://github.com/milvus-io/birdwatcher. By this tool, you can do something hack to milvus.

1. Download the tool from here: https://github.com/milvus-io/birdwatcher/releases
2. Make sure the 2379 port of the etcd container has been exposed
3. In the terminal, start the birdwatcher, enter the command line mode
4. Connect the etcd by command: `connect --etcd [ip]:[port] --rootPath xxx`
5. Remove the missed log path from etcd by this command: `remove binlog --collectionID xxxxxx --partitionID xxxxxx --segmentID xxxxxx --fieldID xxx --logType statslog run`  You can find the collectionID/partitionID/segmentID/fieldID/logType from the full path of the corrupted file.

**Reply by Reidddddd** [reply]
the log path is not missing, it exists, but just its content is corrupted..not readable, couldn't be downloaded

the full path is `insert_log/451813623658484265/451813623658484266/451813623744022898/110/451813623658857039`,  should be binlog or statslog?

**Reply by yhmo** [reply]
"insert_log" is binlog
"stats_log" is statslog

451813623658484265 is collection id
451813623658484266 is partition id
451813623744022898 is segment id
110 is field id

**Reply by Reidddddd** [reply]
This action seems to remove the meta data in etcd.

what will happen on coord(s) and node(s).

is there any watch on these kind of changes?

**Reply by yhmo** [reply]
Yes, the command only remove items from etcd. You can manually delete the corrupted files from minio.

The internal logic is complicated with all kinds of tasks: index task, compaction task, flush task, etc. All these tasks will update the etcd and minio respectively. I don't think it is easy to watch the changes. If you are familar with the milvus source code, then it is possible.

**Reply by Reidddddd** [reply]
i‘m still confused: If only deleting meta in etcd, how does it help make every role back to normal state.

Sound like still needs manual interference.

Manual deleting S3 file is operatable, but how to deal with like `sealed segments` on QueryNode side, also memory data in every coords..

**Reply by yhmo** [reply]
After you delete the path from etcd, the file in minio becomes a garage, and occupies disk space. In fact, it is ok you don't manually delete it from minio, milvus won't read it since the etcd doesn't have its path.

**Reply by Reidddddd** [reply]
「but how to deal with like sealed segments on QueryNode side, also memory data in every coords..」

what about these

**Reply by xiaofan-luan** [reply]
once you cleaned etcd data, restart the cluster could work. because all the memory meta is load from etcd
Btw how could a file on S3 corrupted? any ideas?

**Reply by Reidddddd** [reply]
Underlying is ozone, i'm informed that there are missing blocks and couldn't be recovered

**Comment by Reidddddd** [comment]
Thanks @yhmo @xiaofan-luan

**Comment by yhmo** [comment]
Similar with this discussion: https://github.com/milvus-io/milvus/discussions/37411

@Reidddddd 
Sorry, the command should be:
`remove binlog --collectionID 449576166950553184 --partitionID 453397315907946999 --segmentID 453397315909285711 --fieldID 100 --logType statslog --run`

I forgot to add "--" to "run".
After this command, we will see:
![Screenshot from 2024-11-08 16-51-26](https://github.com/user-attachments/assets/8c2042bd-0e2f-4874-826b-9e1ca450a574)

**Reply by Reidddddd** [reply]
Skimmed the discussion, can i still use logType binlog or delatlog when either of them is unreadable and needs to be cleaned up..

**Reply by yhmo** [reply]
Yes, in the same way.
For example, a delta log missed, you will get error when loading the collection:
`[WARN] [storage/remote_chunk_manager.go:207] ["failed to read object"] [path=files/delta_log/453787183301263392/-1/453787183301463412/453787183301463413]`

Then you can use birdwatcher to remove its path from etcd:
`remove binlog --collectionID 453787183301263392 --partitionID -1 --segmentID 453787183301463412 --logType deltalog --run`
